File: src/core/strings.ts
/**
 * Slugify text: convert to lowercase, remove special chars, replace spaces with hyphens
 */
export function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/\n/g, ' ')
    .replace(/[^a-z0-9\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/^-+|-+$/g, '');
}

/**
 * Remove leading whitespace from multiline strings
 */
export function dedent(text: string): string {
  const lines = text.split('\n');
  const indent = lines.find((line) => line.trim())?.match(/^\s*/)?.[0]?.length ?? 0;
  return lines.map((line) => line.slice(indent)).join('\n');
}

/**
 * Render a string with template variables
 */
export function renderTemplateString(
  template: string,
  variables: Record<string, string | number | boolean>
): string {
  return template.replace(/\{\{(\w+)\}\}/g, (_, key) => {
    const value = variables[key];
    return value !== undefined ? String(value) : '';
  });
}

export function objectToFrontmatter(obj: Record<string, any>): string {
  const entries = Object.entries(obj).map(([key, value]) => {
    if (Array.isArray(value)) {
      return `${key}:\n${value.map((item) => `  - ${item}`).join('\n')}`;
    } else {
      return `${key}: ${value}`;
    }
  });
  return entries.join('\n');
}

File: internal/core/strings.go
package core

import (
	"fmt"
	"regexp"
	"strings"
)

// Slugify converts text to lowercase, removes special chars, replaces spaces with hyphens.
func Slugify(text string) string {
	// Convert to lowercase
	text = strings.ToLower(text)

	// Replace newlines with spaces
	text = strings.ReplaceAll(text, "\n", " ")

	// Remove special characters except alphanumeric, spaces, and hyphens
	re := regexp.MustCompile(`[^a-z0-9\s-]`)
	text = re.ReplaceAllString(text, "")

	// Replace spaces with hyphens
	re = regexp.MustCompile(`\s+`)
	text = re.ReplaceAllString(text, "-")

	// Trim leading/trailing hyphens
	text = strings.Trim(text, "-")

	return text
}

// Dedent removes leading whitespace from multiline strings based on the minimum indent.
func Dedent(text string) string {
	lines := strings.Split(text, "\n")

	// Find minimum indent from non-empty lines
	minIndent := -1
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}

		indent := 0
		for _, ch := range line {
			if ch == ' ' || ch == '\t' {
				indent++
			} else {
				break
			}
		}

		if minIndent == -1 || indent < minIndent {
			minIndent = indent
		}
	}

	if minIndent <= 0 {
		return text
	}

	// Remove minimum indent from each line
	result := make([]string, len(lines))
	for i, line := range lines {
		if len(line) >= minIndent {
			result[i] = line[minIndent:]
		} else {
			result[i] = strings.TrimLeft(line, " \t")
		}
	}

	return strings.Join(result, "\n")
}

// ObjectToFrontmatter converts a map to YAML-style frontmatter.
func ObjectToFrontmatter(obj map[string]any) string {
	var lines []string

	for key, value := range obj {
		switch v := value.(type) {
		case []string:
			lines = append(lines, fmt.Sprintf("%s:", key))
			for _, item := range v {
				lines = append(lines, fmt.Sprintf("  - %s", item))
			}
		case []any:
			lines = append(lines, fmt.Sprintf("%s:", key))
			for _, item := range v {
				lines = append(lines, fmt.Sprintf("  - %v", item))
			}
		default:
			lines = append(lines, fmt.Sprintf("%s: %v", key, v))
		}
	}

	return strings.Join(lines, "\n")
}
